File: Python/2289. Steps to Make Array Non-decreasing.py
from typing import List
class Solution:
    def totalSteps(self, nums: List[int]) -> int:
        st = []
        n = len(nums)
        ans = 0

        st.append((nums[n - 1], 0))

        for i in range(n - 2, -1, -1):
            cnt = 0

            while st and nums[i] > st[-1][0]:
                cnt = max(cnt + 1, st.pop()[1])

            ans = max(ans, cnt)
            st.append((nums[i], cnt))

        return ans

File: C++/2289. Steps to Make Array Non-decreasing.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int totalSteps(vector<int>& nums) {
        stack<pair<int, int>> st;
        int n = nums.size();
        int ans = 0;

        st.push({ nums[n - 1],0 });

        for (int i = n - 2;i >= 0;i--) {
            int cnt = 0;
            while (!st.empty() && nums[i] > st.top().first) {
                cnt = max(cnt + 1, st.top().second);
                st.pop();
            }
            ans = max(ans, cnt);
            st.push({ nums[i],cnt });
        }
        return ans;
    }
};
